### DISCUSSION TITLE
Video Codec Hardware Requirements and Support for Linux

### DISCUSSION THREAD

**Discussion by ToruHironaka** [rustdesk/rustdesk#2925] [General]
I would like to know the hardware requirements for rustdesk's video codec options (Auto, VP9, H264, H265. Please see the attached image below). I think ruskdesk requires NVIDIA GPU GTX/RTX 600 or above for the video codec options. The video codec option appears on rustdesk display setting if both client and host PCs' GPUs are NVIDIA GPU GTX/RTX 600 or above and they are both Windows 10. Otherwise, the video codec option did not appear on the display setting. 

I would like to also know the video codec option is available on Linux platform, because the video codec option did not appear on rustdesk display setting if either one of PCs is Linux, even though their GPUs are NVIDIA GPU GTX/RTX 600 or above. 

![image](https://user-images.githubusercontent.com/6956496/214147982-929f5d1a-63fc-4e4e-8ab7-885c062add17.png)

**Comment by 21pages** [comment]
It depends on decoding support of local and  encoding support of remote. The support state will be checked every time the program is started.

**Comment by ToruHironaka** [comment]
My rsutdesk host showed the message below, so the rustdesk host seemed to find the codecs VP9, H265, and H265. But, my rustdesk client still did not show auto or other hardware codec options in the setting. 

video_codec_state: MessageField(Some(VideoCodecState { score_vpx: 90, score_h264: 94, score_h265: 95, prefer: Auto

![image](https://user-images.githubusercontent.com/6956496/214931480-fbfbcb2a-bd5e-48ca-a282-de4ad175a3c9.png)

My rustdesk client's video codec was using hardware codec VP9.

![image](https://user-images.githubusercontent.com/6956496/214931311-39dbc9a0-656c-4049-98c8-07d394e748cb.png)

 My other remote desktop client "Nomachine" was using codec H264. 

![image](https://user-images.githubusercontent.com/6956496/214931021-868aeac3-8c0f-4ce0-aebc-a6fbb843c322.png)

So, both host and client could detect both VP9 and H264. I ran rustdesk with "--feature hwcodec" option, which did not work either. I also ran "cargo update -p hwcodec" to update rustdesk hwcodec, and the update was successful. What am I missing? or Is rustdesk out of support of hardware codec on Linux OS? 

![image](https://user-images.githubusercontent.com/6956496/214933239-0e031182-a124-4b29-b516-ec376e300490.png)

**Comment by ToruHironaka** [comment]
Well, I figured it out. Rustdesk 1.2.0 required nv-codec-headers and ffmpeg with NVIDIA GPU hardware acceleration. I followed the attached instruction below to install nv-codec-headers and ffmpeg, and ran rustdesk with "--feature hwcodec" option (maybe, I do not need this option), and then my rustdesk client showed me hardware code options (Auto, VP9, H264, and H265). 

![image](https://user-images.githubusercontent.com/6956496/214942937-4fef3f1f-0896-49cf-8eed-b65369baf692.png)


[Using_FFmpeg_with_NVIDIA_GPU_Hardware_Acceleration.pdf](https://github.com/rustdesk/rustdesk/files/10513128/Using_FFmpeg_with_NVIDIA_GPU_Hardware_Acceleration.pdf)

**Comment by 21pages** [comment]
> video_codec_state: MessageField(Some(VideoCodecState { score_vpx: 90, score_h264: 94, score_h265: 95, prefer: Auto

This is the decoding supported by local.
